allenact/embodiedai/aux_losses/losses.py

```diff
@@ -0,0 +1,599 @@
+            preds.view((num_steps - 1) * num_sampler, -1),  # ((T-1)*B, A)
+            actions.flatten(),  #  ((T-1)*B,)
+        )
+        loss = loss.view(num_steps - 1, num_sampler)  # (T-1, B)
+
+        def vanilla_valid_losses(loss, num_sampler, end_locs_batch):
+            ##  this is just used to verify the vectorized version works correctly.
+            ##  not used for experimentation
+            valid_losses = []
+            for i in range(num_sampler):
+                end_locs = end_locs_batch[i]
+                for j in range(len(end_locs)):
+                    if j == 0:
+                        start_loc = 0
+                    else:
+                        start_loc = end_locs[j - 1] + 1
+                    end_loc = end_locs[j]
+                    if end_loc - start_loc <= 0:  # the episode only 1-step
+                        continue
+                    valid_losses.append(loss[start_loc:end_loc, i])
+
+            if len(valid_losses) == 0:
+                valid_losses = torch.zeros(1, dtype=torch.float).to(loss)
+            else:
+                valid_losses = torch.cat(valid_losses)  # (sum m, )
+            return valid_losses
+
+        # valid_losses = masks[1:] * loss # (T-1, B)
+        # valid_losses0 = vanilla_valid_losses(loss, num_sampler, end_locs_batch)
+        # assert valid_losses0.sum() == valid_losses.sum()
+
+        num_valid_losses = torch.count_nonzero(masks[1:])
+        if num_valid_losses < self.subsample_min_num:  # don't subsample
+            subsample_rate = 1.0
+        else:
+            subsample_rate = self.subsample_rate
+
+        loss_masks = masks[1:] * _bernoulli_subsample_mask_like(
+            masks[1:], subsample_rate
+        )
+        num_valid_losses = torch.count_nonzero(loss_masks)
+        avg_loss = (loss * loss_masks).sum() / torch.clamp(num_valid_losses, min=1.0)
+
+        return (
+            avg_loss,
+            {"total": cast(torch.Tensor, avg_loss).item(),},
+        )
+
+
+class TemporalDistanceLoss(AuxiliaryLoss):
+    """
+    Auxiliary task of Temporal Distance
+        from Auxiliary Tasks Speed Up Learning PointGoal Navigation (Ye, 2020) https://arxiv.org/abs/2007.04561
+    """
+
+    UUID = "TempDist"
+
+    def __init__(self, num_pairs: int = 8, epsiode_len_min: int = 5, *args, **kwargs):
+        super().__init__(auxiliary_uuid=self.UUID, *args, **kwargs)
+        self.num_pairs = num_pairs
+        self.epsiode_len_min = float(epsiode_len_min)
+
+    def get_aux_loss(
+        self,
+        aux_model: nn.Module,
+        observations: ObservationType,
+        obs_embeds: torch.FloatTensor,
+        actions: torch.FloatTensor,
+        beliefs: torch.FloatTensor,
+        masks: torch.FloatTensor,
+        *args,
+        **kwargs
+    ):
+        ## we discard the last action in the batch
+        num_steps, num_sampler = actions.shape  # T, B
+        actions = cast(torch.LongTensor, actions)
+        actions = actions[:-1]  # (T-1, B)
```

Comment:
LGTM alert says that `actions` isn't used anywhere below so this line can probably be removed.